src/librustc/middle/infer/combine.rs

```diff
@@ -278,7 +278,7 @@ impl<'a, 'tcx> CombineFields<'a, 'tcx> {
         };
         let u = ty.fold_with(&mut generalize);
         if generalize.cycle_detected {
-            Err(ty::terr_cyclic_ty)
```

Comment:
would you be okay with just TypeError::Variant? I think it is slightly less verbose and just as clear.